astropy/units/format/utils.py

```diff
@@ -4,27 +4,41 @@
 Utilities shared by the different formats.
 """
 
+from __future__ import annotations
+
 import warnings
+from typing import TYPE_CHECKING
 
 from astropy.units.utils import maybe_simple_fraction
 from astropy.utils.misc import did_you_mean
 
+if TYPE_CHECKING:
+    from collections.abc import Callable, Container, Mapping, Sequence
+    from numbers import Real
+    from typing import TypeVar
+
+    from astropy.units import UnitBase
+
+    T = TypeVar("T")
+
 
-def get_grouped_by_powers(bases, powers):
+def get_grouped_by_powers(
+    bases: Sequence[T], powers: Sequence[int]
+) -> tuple[list[tuple[T, int]], list[tuple[T, int]]]:
     """
     Groups the powers and bases in the given
     `~astropy.units.CompositeUnit` into positive powers and
     negative powers for easy display on either side of a solidus.
 
     Parameters
     ----------
-    bases : list of `astropy.units.UnitBase` instances
+    bases : `~collections.abc.Sequence` of `astropy.units.UnitBase`
 
-    powers : list of int
+    powers : `~collections.abc.Sequence` of `int`
 
     Returns
     -------
-    positives, negatives : tuple of lists
+    positives, negatives : `tuple` of `list`
```

Comment:
I think this docstring started off incorrect. Numpy docstrings, on a collection of Parameters or Returns refer to the elements of each, not the collection. It's `*args : ElementType` not `*args : tuple[ElementType]`, so this should be

```suggestion
    positives, negatives : `list` of `tuple` of |Unit| and power
```

Or something like that.

Reply:
Are we sure we need the back quotes? I thought numpydoc did that automatically - it also is distracting for people who read the docstring in ipython (maybe not so relevant here, but perhaps just in an editor). So just "list of tuple of |Unit| and power".